### PR TITLE
JP-2500 Allow cube build to accept a tangent point, position angle and number of spaxels in x, y cube dimensions

### LINKED ISSUE #6697: Specify tangent projection details for IFU spectral cubes
_Issue [JP-2500](https://jira.stsci.edu/browse/JP-2500) was created on JIRA by [Karl Gordon](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kgordon):_

A useful capability would be to be able to specify the tangent projection details (center ra, dec and rotation angle) for the IFU spectral cubes.  The goal here is is to be able to build all the IFU spectral cubes (e.g., per grating setting) for NIRSpec and MIRI on the same final tangent projection.  This enables science investigations on the same region of the sky (e.g., pixel in tangent projection) without having to interpolate the different IFU cubes to the same projection.  

This would be a user settable option for cube_build with the goal of this being done offline by individual users re-running cube_build.

I briefly discussed this with [Jane Morrison](https://jira.stsci.edu/secure/ViewProfile.jspa?name=morrison) who indicated this capability was possible with the existing cube_build implementation.

### PR DESCRIPTION
**Pull request by jemorrison**
Resolves [JP-2500](https://jira.stsci.edu/browse/JP-2500)

Closes #6697

This PR addresses allows the user to specify the tangent point (ra_center, dec_center), position angle (cube_pa) and size of cube in x cube dimension (nspax_xi) and y cube dimension (nspax_eta). 
These are all option to be run offline not part of standard processing. 

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
      Results of regression test:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/955/
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by jemorrison** [comment]
@drlaw1558 Test this out and let me know what you think.
I am assuming a user provided ra_center or dec_center is just a smallish change to the current values. If a user gives a very large value (say at the edge of the cube) that becomes the tangent point. Cube build will try and create a cube with the center of the cube at this location. So there can be empty space where no data exist. This can be made smaller by sending in an nspax_xi or nspax_eta but if the tangent point at the edge there still will be empty space. But I am also assuming that most people will not do that and the ra,dec centers are moved just a little bit. 
If this PR looks corrrect I will update the docs and add the new terms

File: jwst/cube_build/ifu_cube.py [hunk 1]
```diff
@@ -255,8 +261,30 @@ def set_geometry(self, corner_a, corner_b, lambda_min, lambda_max):
         # we have angles in degrees
         ra_ave = circmean(ravalues * u.deg).value
 
-        self.crval1 = ra_ave
-        self.crval2 = dec_ave
+        if self.ra_center is not None:
+            # check that user provided ra center are within the already
+            # determined min and max values.
```
Comment (drlaw1558) [inline jwst/cube_build/ifu_cube.py:266]: Do we need to require this?  I'm thinking ahead to some programs that tile larger galaxies, and the desired cube center could be outside the footprint of any individual observation.  I think the consequence would be just that the cubes get absurdly big right?

Comment (jemorrison) [reply jwst/cube_build/ifu_cube.py:266]: Ah ok. It was just a check for users not entering incorrect values. But I had not thought of that case. I will remove it

**Comment by drlaw1558** [comment]
Hm, I just tested moving the ra/dec center and for some reason the flux values in the cube are now all different.  I'll see if I can sort out why.

**Review by drlaw1558 — APPROVED** [review]
LGTM; I tested each of these parameters and combinations thereof, and the results looked as expected.  Needs a change log entry though.

**Comment by jemorrison** [comment]
I will update the documentation now. I just wanted to make sure the code looked correct before I did that.

File: CHANGES.rst [hunk 2]
```diff
@@ -77,6 +77,9 @@ cube_build
 
 - Fix a bug with memory allocation in C extensions when weighting=emsm. [#7847]
 
+- Add options to set ra,dec tangent projection center, position angle and size of cube
+  in tangent plane.[#7882]
```
Comment (hbushouse) [inline CHANGES.rst:81]: need a whitespace between the period and opening bracket of the PR number

File: docs/jwst/cube_build/arguments.rst [hunk 3]
```diff
@@ -76,6 +76,21 @@ The following arguments control the size and sampling characteristics of the out
...
+  Right ascension center, in decimal degrees, of the IFU cube that defines the location of xi/eta tangent plane projection origin.
+
+``dec_center``
+  Declination center, in decimal degrees, of the IFU cube that defines the location of xi/eta tangent plane projection origin.
+
+``cube_pa``
+  The position angle of the IFU cube in decimal degrees (E from N).
+
+``nspax_xi``
+  The odd integer number of spaxels to use in the xi dimension of the tangent plane.
```
Comment (hbushouse) [inline docs/jwst/cube_build/arguments.rst:89]: I assume savy users know what "xi" and "eta" are?

File: jwst/cube_build/cube_build_step.py [hunk 4]
```diff
@@ -108,6 +113,19 @@ def process(self, input):
         if self.roiw != 0.0:
             self.log.info(f'Input Wave ROI size {self.roiw}')
 
+        # check that if self.nspax_xi or self.nspax_eta is provided they must be even numbers
```
Comment (hbushouse) [inline jwst/cube_build/cube_build_step.py:116]: this comment says *even*, but the log message below says *odd*

File: jwst/cube_build/ifu_cube.py [hunk 5]
```diff
@@ -284,6 +297,13 @@ def set_geometry(self, corner_a, corner_b, lambda_min, lambda_max):
         na = math.ceil(xilimit / self.cdelt1) + 1
         nb = math.ceil(etalimit / self.cdelt2) + 1
 
+        # if the user set the nspax_xi or nspax_eta then redefine na, nb
+        # it is assumed that both values are ODD numbers
```
Comment (hbushouse) [inline jwst/cube_build/ifu_cube.py:301]: So, given the confusion above, do you really mean ODD here or not?

Comment (jemorrison) [reply jwst/cube_build/ifu_cube.py:301]: My bad. I had not cleaned that up yet. I originally thought the numbers needed to be even until I tested it and realized I needed them to odd.

**Comment by jemorrison** [comment]
@drlaw1558 @hbushouse 
Howard mentions in a comment about the use of xi and eta. I always think of the tangent plane projection in those coordinates but I realize we never have mentioned xi, eta in the documentation. Should I add a line or two describing that xi, eta are the tangent plane coordinates the cube is build on in the Terminology section of the description or should I change nspax_xi to nspax_x and nspax_eta to nspax_y and refer to those values as the x,y values in the tangent project ? And remove other references to xi and eta ?

**Comment by drlaw1558** [comment]
@jemorrison Hm, https://github.com/spacetelescope/jwst/pull/7783 used the convention scalexy, so by analogy nspax_x and nspax_y would probably make sense.  xi/eta and x/y aren't the same values so we need to keep both, but should be interchangeable from the standpoint of numbers of spaxels.  That way we wouldn't have to describe xi/eta in anything other than inline documentation as this is just used internally as an intermediate step in the tangent plane projection.

**Review by hbushouse — APPROVED** [review]
Looks good now.

**Comment by hbushouse** [comment]
I started a regtest run here: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/932
Theoretically, since this update only involves params used by a user in off-line processing, it should have no effect at all on the regtests. Although perhaps someone should add a new test that utilizes these params.

**Comment by jemorrison** [comment]
I can add a regression test that sets the size of the cube. Cube build requires a full image to run so I could not figure out a way to test that in a unit test. But I can add a regression tests. I am hesitant to do that right now because of all the regression test failure to run.

File: JenkinsfileRT [hunk 6]
```diff
@@ -43,7 +43,7 @@ jobconfig.publish_env_on_success_only = true
 jobconfig.publish_env_filter = "spacetelescope/master"
 
 // Define python version for conda
-python_version = "3.11"
+python_version = "3.9"
```
Comment (hbushouse) [inline JenkinsfileRT:46]: A real fix for the regtest server issues has been applied to jwst/master. So please remove this change from your PR and rebase your PR branch to pull in those changes from master. You should be able to run a regtest on the PR branch then.

**Comment by jemorrison** [comment]
Results of regression test:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/955/

@hbushouse is it ok to merge this - or should I wait since Build 10 is being made now ?

**Pull request merged**